src/useMutation.ts

```diff
@@ -32,18 +37,155 @@ export type MutationFn<TData, TVariables> = (
+  error: undefined,
+  hasError: false,
+  loading: false,
+});
+
+export function useMutation<TData, TVariables = OperationVariables>(
   mutation: DocumentNode,
-  {
-    client: overrideClient,
-    ...baseOptions
-  }: MutationHookOptions<TData, TVariables, TCache> = {}
-): MutationFn<TData, TVariables> {
-  const client = useApolloClient(overrideClient);
-
-  return options => client.mutate({ mutation, ...baseOptions, ...options });
+  baseOptions: MutationHookOptions<TData, TVariables> = {}
+): [MutationFn<TData, TVariables>, MutationResult<TData>] {
+  const client = useApolloClient(baseOptions.client);
+  const [result, setResult] = React.useState<MutationResult<TData>>(
+    getInitialState
+  );
+
+  const { throwMode = 'async', ...options } = baseOptions;
+
+  React.useEffect(
+    () => {
+      if (throwMode === 'sync' && result.error) {
+        throw result.error;
+      }
+    },
+    [throwMode, result.error]
+  );
+
+  // reset state if client instance changes
+  React.useLayoutEffect(
+    () => {
+      setResult(getInitialState);
+    },
+    [client]
+  );
+
+  const { generateNewMutationId, isMostRecentMutation } = useMutationTracking();
+
+  const onMutationStart = () => {
+    if (!result.loading) {
+      // A hack to get rid React warnings during tests.
+      actHack(() => {
+        setResult({
+          called: true,
+          data: undefined,
+          error: undefined,
+          hasError: false,
+          loading: true,
+        });
+      });
+    }
+  };
+
+  const onMutationError = (error: ApolloError, mutationId: number) => {
+    if (isMostRecentMutation(mutationId)) {
+      // A hack to get rid React warnings during tests.
+      actHack(() => {
+        setResult(prev => ({
+          ...prev,
+          error: new ApolloMutationError(error, mutation, baseOptions),
+          hasError: true,
+          loading: false,
+        }));
+      });
+    }
+  };
+
+  const onMutationCompleted = (
+    response: ExecutionResult<TData>,
+    mutationId: number
+  ) => {
+    const { data, errors } = response;
+    if (errors && errors.length > 0) {
+      onMutationError(new ApolloError({ graphQLErrors: errors }), mutationId);
+      return;
+    }
+
+    if (isMostRecentMutation(mutationId)) {
+      // A hack to get rid React warnings during tests.
+      actHack(() => {
+        setResult(prev => ({
+          ...prev,
+          data,
+          loading: false,
+        }));
+      });
+    }
+  };
+
+  const runMutation = React.useCallback(
+    async (mutateOptions: MutationHookOptions<TData, TVariables> = {}) => {
+      onMutationStart();
+      const mutationId = generateNewMutationId();
+
+      try {
+        // merge together variables from baseOptions (if specified)
+        // and the execution
+        const mutateVariables = options.variables
+          ? { ...mutateOptions.variables, ...options.variables }
+          : mutateOptions.variables;
+
+        const response = await client.mutate({
```

Comment:
I would hardly recommend not to use `async-await` syntax. It increases amount of [output code](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=KYDwDg9gTgLgBAMwK4DsDGMCWEVyQZ2AFkkYBDLHACgCg44BbUi7FAGjrgCMzCB5MJRT44AXjgBvAL40AlJM5oc-eGgA2mYCnjiCwAIKQ1aiAGENWmFR79BrfADp1m7bIDci5fADaUYPiQ1GDY4QhgAJX9AmABdMThIsgwHPQBlchhgWnp6AHNgGABJFEwsMjV0imBOdxpPYXgJOBgACygIAHciCAATYHiAcl4AT3QBkIdJiDtlOCl4m2ABIXwPTkTkvQBRBARgDGycqnlRAD4FHJzMBDgqVvau3v7RF7gB_FG0AbgAMh-4PwBIIOYBQdpQeQSTiXej3ToAqLA0HgjwwubQuYcS7eOGPPohQHREFg6AxGpregAekpCLCoQy_WucGcljgmAaZHQ_TQLU5-Xw62ASRgKUIABkyMMIKQdnsDhjjmJzlC0WFIkCrPkiiUyhUGbVLlIsTlvCztGT6LV6ipJHB8ihQVUAHLALrMISFHohTD4boqSJoSwkDKsObxPTBlg4AAqUCSAGt2bljhTmV44DhI0JKrB4oqzhcrjcqABCQnAkxkHpJyEYqk0_RwXloePNCB2goAzA9BJCjBwDpkKAlFC5EQ9JBQJPNfwwRx1uDCgASCao-eVC_oasRVhVaJyaHKamAPQAXM0oEhgMb93AehQyOfUH0EOyTzf98joE-UC-317NxyXl8C2EkoHPBBykID80UratR3PGBL2vQCpANGE0NRHIZHoKRUyUBoMxQLNWFA8E8y_KAQiYEMcE9E4N0uJkqB9P0In2IN3VYKgaKjFB6NrNFqTgRtm1bGB2y1Lsew2eBB2HJNx0nadMhUec0WXVd10LVUCnVaIqDAPwADclVuPdb0mBwjOAYyYJhSjzwdDoRKMEwSJwMjoCoSjqK4nAQkWZZ7FkezLmArzwIvK8wpyOCkwgqCUNvND0MNNL0Vw_D00zfyUFMCAGDAY9Mh7cRDnoQFIGEZKcl4j0ehqMyLLTQimnvcgQkokR5nEKrlGALD6GY7rfn-bqHGPUdWjgc4AAZBJhXLaJQSKqGc1yIGMCA1qaXI4zAFoAEUxUi_Bz1GtC_JW-ihpyPwYEnFA7pkDFmNYiB_Q47QPJQHi8oEnScmE0SEzbDt4CnGS-zkocRzHO9lNHGc1IcBdNJbNcGKBy5tw1QyTLMqgWrRKybLswD6A6shYvoeKEMQJLYtSu7cIynC5mywjL2IvL4lk0VgFMI8eExjERnQW56qWGZhGxkmiN-nMrAy-gCJterWE9eJ7UdTIXTdG6ehTOo0SQ4YceBmkGFBfI2y1FpQTgYyh0wMguGPEQEHaBhuF4GWVluJl8DAfZrk0HpZEA4TOR7Vp-lAfZSFYQD1fgaWADVXfdz34mmFYHBdqcc_8Sm4AAfltKzpeC5RC-zj3_AmKZZccIu3cbnqy_PGvW_r4vO9TNE09parCHiMhB1KZkLG0BxpeJsvNYCsurPz-xabgav3QDjey_bkvzsYHes4Hz3matW9lr4gqipKk8qH6mrrr427AIep7R4G1mw0PGAeVuMiRaaJr5CDWsiF-DVVZFluLibofQxCvCGB8MYwD9y4jgMiH-uF34FE_tIH-HNcIflNLPYIx8VqBX9rXYQIQIBcAAFbRggAAaWAMMaw1DW6yAtHAS-CJHrDjgL4VAv0CQ7hiB4GQQA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=false&presets=react%2Ctypescript%2Cenv&prettier=false&targets=&version=7.3.4), makes it hard to debug in browser devtools and requires `regeneratorRuntime` polyfill.

Reply:
Thanks, async-await is a goner :)